agent/recorder.go

```diff
@@ -167,19 +188,53 @@ func (r *SpanRecorder) SendSpans() (error, bool) {
 	}
 
 	if payloadSent {
-		r.okSend++
+		atomic.AddInt64(&r.stats.sendSpansOk, 1)
+		atomic.AddInt64(&r.stats.spansSent, int64(len(spans)))
 	} else {
-		r.koSend++
+		atomic.AddInt64(&r.stats.sendSpansKo, 1)
+		atomic.AddInt64(&r.stats.spansNotSent, int64(len(spans)))
 	}
 
 	return lastError, shouldExit
 }
 
+// Stop recorder
+func (r *SpanRecorder) Stop() {
+	if r.debugMode {
+		r.logger.Println("Scope recorder is stopping gracefully...")
+	}
+	r.t.Kill(nil)
+	_ = r.t.Wait()
+	if r.hasSpans() {
+		err, _ := r.SendSpans()
+		if err != nil {
+			r.logger.Printf("error sending spans: %v\n", err)
+		}
```

Comment:
Well there is a corner case, the loop func is not locking (which is good), so if we Stop the recorder, the `dying` channel call the SendSpans (with the http request), meanwhile we can receive a span, and because the tomb is still alive (because the loop has not ended) the span is added to the span slice. So then the loop finish and the Alive is set to false and the Wait in the stop method resumes.

So this will handle that corner case and ensure all the received spans are sent.

Reply:
https://godoc.org/gopkg.in/tomb.v2#Tomb.Alive

```
Alive returns true if the tomb is not in a dying or dead state.
```

So we would reject the received span while in dying state.

Reply:
Moreover, the Kill and Alive locks, when the tomb is dying the RecordSpan method will wait until the dying channel gets closed.

![image](https://user-images.githubusercontent.com/69803/74225004-31b82080-4cba-11ea-92f3-dd12174dfa28.png)

I'll remove the unnecessary code